jenkins/Jenkinsfile

```diff
@@ -1,203 +1,167 @@
-    // require a whole system to be running.
-    stage('Deploy to test') {
-      steps {
-      sh './gradlew deployToTestWindowsLocal -Pdeploy_directory=\"/Users/matthew/Downloads/apache-tomcat-9.0.37/webapps\"'
-      // pipenv needs to be installed and on the path for this to work.
-      sh 'PIPENV_IGNORE_VIRTUALENVS=1 pipenv install'
+  // Runs an analysis of the code, looking for any
+  // patterns that suggest potential bugs.
+  stage('Static Analysis') {
+      sh './gradlew sonarqube'
+      // wait for sonarqube to finish its analysis
+      sleep 5
+      sh './gradlew checkQualityGate'
+  }
 
-      // Wait here until the server tells us it's up and listening
-      sh './gradlew waitForHeartBeat'
-      }
-    }
 
+  // Move the binary over to the test environment and
+  // get it running, in preparation for tests that
+  // require a whole system to be running.
+  stage('Deploy to test') {
+    sh './gradlew deployToTestWindowsLocal -Pdeploy_directory=\"/Users/matthew/Downloads/apache-tomcat-9.0.37/webapps\"'
+    // pipenv needs to be installed and on the path for this to work.
+    sh 'PIPENV_IGNORE_VIRTUALENVS=1 pipenv install'
 
-    // Run the tests which investigate the functioning of the API.
-    stage('API Tests') {
-      steps {
-        sh './gradlew runApiTests'
-      }
-      post {
-        always {
-          junit 'build/test-results/api_tests/*.xml'
-        }
-      }
-    }
+    // Wait here until the server tells us it's up and listening
+    sh './gradlew waitForHeartBeat'
+  }
 
-    // We use a BDD framework for some UI tests, Behave, because Python rules
-    // when it comes to experimentation with UI tests.  You can try things and see how they work out.
-    // this set of BDD tests does require a running system.
-    // BDD at the UI level is just to ensure that basic capabilities work,
-    // not that every little detail of UI functionality is correct.  For
-    // that purpose, see the following stage, "UI Tests"
-    stage('UI BDD Tests') {
-      steps {
-        sh './gradlew runBehaveTests'
-        sh './gradlew generateCucumberReport'
-      }
-      post {
-        always {
-          junit 'build/test-results/bdd_ui/*.xml'
-        }
-      }
-    }
 
-    // This set of tests investigates the functionality of the UI.
-    // Note that this is separate fom the UI BDD Tests, which
-    // only focuses on essential capability and therefore only
-    // covers a small subset of the possibilities of UI behavior.
-    stage('UI Tests') {
-        steps {
-            sh 'cd src/ui_tests/java && ./gradlew clean test'
-        }
-        post {
-            always {
-                junit 'src/ui_tests/java/build/test-results/test/*.xml'
-            }
-        }
-    }
+  // Run the tests which investigate the functioning of the API.
+  stage('API Tests') {
+    result = sh(script: './gradlew runApiTests', returnStatus: true)
+    results["API Tests"]["status"] = (result == 0) ? 2 : 1
```

Comment:
Wouldn't mind a little note here about what 2 and 1 mean

Reply:
This is a good point. 2 is unintuitively passing and 1 is failing, which is why we need this ternary everywhere